framework/src/modules/chain/synchronizer/synchronizer.js

```diff
@@ -14,47 +14,175 @@
-const FastChainSwitchingMechanism = require('./fast_chain_switching_mechanism');
+const util = require('util');
+const assert = require('assert');
+
+const {
+	verifySignature,
+	verifyVersion,
+	verifyReward,
+	verifyId,
+	verifyPayload,
+} = require('../blocks/verify');
 
 class Synchronizer {
-	constructor() {
-		this.currentStrategy = null;
-		this.blockSynchronizationMechanism = new BlockSynchronizationMechanism();
-		this.fastChainSwitchingMechanism = new FastChainSwitchingMechanism();
+	constructor({
+		storage,
+		logger,
+		blockReward,
+		exceptions,
+		maxTransactionsPerBlock,
+		maxPayloadLength,
+	}) {
+		this.storage = storage;
+		this.logger = logger;
+		this.blockReward = blockReward;
+		this.exceptions = exceptions;
+
+		this.constants = {
+			maxPayloadLength,
+			maxTransactionsPerBlock,
+		};
+
+		this.mechanisms = [];
+	}
+
+	/**
+	 * Register a sync mechanism with synchronizer
+	 * It must have "isValidFor" and "run" interface to call upon
+	 *
+	 * "isValidFor" must return true/false to match the sync mechanism
+	 * "isValidFor" can throw error to abort the synchronization
+	 *
+	 * "run" must initiate the synchronization
+	 * "run" must keep track of its state internally
+	 *
+	 * @param {Object} mechanism - Mechanism to register
+	 */
+	register(mechanism) {
+		assert(
+			util.types.isAsyncFunction(mechanism.isValidFor),
+			'Sync mechanism must have "isValidFor" async interface'
+		);
+		assert(
+			util.types.isAsyncFunction(mechanism.run),
+			'Sync mechanism must have "run" async interface'
+		);
+
+		// Check the property isActive, it can be own property or a getter
+		assert(
+			mechanism.isActive !== undefined,
+			'Sync mechanism must have "isActive" interface'
+		);
+
+		this.mechanisms.push(mechanism);
 	}
 
 	/**
 	 * Start the syncing mechanism
+	 * @param {Object} receivedBlock - The block you received from network, used to choose sync mechanism
 	 * @return {*}
 	 */
-	run() {
-		if (this.currentStrategy && this.currentStrategy.isActive()) {
-			throw new Error('Blocks Sychronizer is already running');
+	async run(receivedBlock) {
+		if (this.activeMechanism) {
+			throw new Error(
+				`Blocks Sychronizer with ${
+					this.activeMechanism.constructor.name
+				} is already running`
+			);
+		}
+
+		const lastBlock = await this.storage.entities.Block.getLastBlock();
+
+		// Moving to a Different Chain
+		// 1. Step: Validate new tip of chain
+		const result = this._verifyBlockBeforeSync(lastBlock, receivedBlock);
+		if (!result.verified) {
+			throw Error(
+				`Block verification for chain synchronization failed with errors: ${result.errors.join()}`
+			);
 		}
 
-		this.currentStrategy = this._determineStrategy();
+		// Choose the right mechanism to sync
+		const validMechanism = await this._determineSyncMechanism(receivedBlock);
 
-		return this.currentStrategy.run();
+		if (!validMechanism) {
+			return this.logger.info(
+				"Can't determine sync mechanism at the moment for block",
```

Comment:
```suggestion
				"Sync mechanism could not be determined for the given block",
```

I think this way is clearer